/* -------------------------------------------------------------------------- */
/* Copyright 2002-2025, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/* -------------------------------------------------------------------------- */
syntax = "proto3";

import "shared.proto";

package one.image;

option go_package = "github.com/OpenNebula/one/src/oca/go/src/goca/api/image";

message AllocateRequest
{
  string session_id        = 1;
  string template          = 2;
  int32 ds_id              = 3;
  bool skip_capacity_check = 4;
}

message DeleteRequest
{
  string session_id = 1;
  int32 oid         = 2;
  bool force        = 3;
}

message InfoRequest
{
  string session_id = 1;
  int32 oid         = 2;
  bool decrypt      = 3;
}

message UpdateRequest
{
  string session_id = 1;
  int32 oid         = 2;
  string template   = 3;
  int32 append      = 4;
}

message RenameRequest
{
  string session_id = 1;
  int32 oid         = 2;
  string name       = 3;
}

message ChmodRequest
{
  string session_id   = 1;
  int32 oid           = 2;
  sint32 user_use     = 3;
  sint32 user_manage  = 4;
  sint32 user_admin   = 5;
  sint32 group_use    = 6;
  sint32 group_manage = 7;
  sint32 group_admin  = 8;
  sint32 other_use    = 9;
  sint32 other_manage = 10;
  sint32 other_admin  = 11;
}

message ChownRequest
{
  string session_id = 1;
  int32 oid         = 2;
  sint32 user_id    = 3;
  sint32 group_id   = 4;
}

message LockRequest
{
  string session_id = 1;
  int32 oid         = 2;
  int32 level       = 3;
  bool test         = 4;
}

message UnlockRequest
{
  string session_id = 1;
  int32 oid         = 2;
}

message CloneRequest
{
  string session_id = 1;
  int32 oid         = 2;
  string name       = 3;
  int32 ds_id       = 4;
}

message EnableRequest
{
  string session_id = 1;
  int32 oid         = 2;
  bool enable       = 3;
}

message PersistentRequest
{
  string session_id = 1;
  int32 oid         = 2;
  bool persistent   = 3;
}

message ChtypeRequest
{
  string session_id = 1;
  int32 oid         = 2;
  string type       = 3;
}

message SnapshotDeleteRequest
{
  string session_id = 1;
  int32 oid         = 2;
  int32 snapshot_id = 3;
}

message SnapshotRevertRequest
{
  string session_id = 1;
  int32 oid         = 2;
  int32 snapshot_id = 3;
}

message SnapshotFlattenRequest
{
  string session_id = 1;
  int32 oid         = 2;
  int32 snapshot_id = 3;
}

message ResizeRequest
{
  string session_id = 1;
  int32 oid         = 2;
  string size       = 3; // New size in MiB as integer string, must be > current size
}

message RestoreRequest
{
  string session_id = 1;
  int32 oid         = 2;
  int32 ds_id       = 3;
  string opt_tmpl   = 4;
}

message PoolInfoRequest
{
  string session_id  = 1;
  sint32 filter_flag = 2;
  sint32 start       = 3;
  sint32 end         = 4;
}

service ImageService
{
  rpc Allocate (one.image.AllocateRequest) returns (one.ResponseID);

  rpc Delete (one.image.DeleteRequest) returns (one.ResponseID);

  rpc Info (one.image.InfoRequest) returns (one.ResponseXML);

  rpc Update (one.image.UpdateRequest) returns (one.ResponseID);

  rpc Rename (one.image.RenameRequest) returns (one.ResponseID);

  rpc Chmod (one.image.ChmodRequest) returns (one.ResponseID);

  rpc Chown (one.image.ChownRequest) returns (one.ResponseID);

  rpc Lock (one.image.LockRequest) returns (one.ResponseID);

  rpc Unlock (one.image.UnlockRequest) returns (one.ResponseID);

  rpc Clone (one.image.CloneRequest) returns (one.ResponseID);

  rpc Enable (one.image.EnableRequest) returns (one.ResponseID);

  rpc Persistent (one.image.PersistentRequest) returns (one.ResponseID);

  rpc Chtype (one.image.ChtypeRequest) returns (one.ResponseID);

  rpc SnapshotDelete (one.image.SnapshotDeleteRequest) returns (one.ResponseID);

  rpc SnapshotRevert (one.image.SnapshotRevertRequest) returns (one.ResponseID);

  rpc SnapshotFlatten (one.image.SnapshotFlattenRequest) returns (one.ResponseID);

  rpc Resize (one.image.ResizeRequest) returns (one.ResponseID);

  rpc Restore (one.image.RestoreRequest) returns (one.ResponseXML);

  rpc PoolInfo (one.image.PoolInfoRequest) returns (one.ResponseXML);
}
